### ISSUE TITLE
Performance issues / memory core dumps with big files

### ISSUE THREAD

**Issue by melvinroest** [rubensworks/jsonld-streaming-parser.js#65]
I made a program `node test.js `. I ran this with `12.18.0` and with `14.4.0` for good measure.

It looks as follows:

```
const fs = require('fs');
const lib = require('./index');
const path = "./authorities-geografikum_lds_20200213.jsonld"; //downloadable from https://data.dnb.de/opendata/authorities-geografikum_lds.jsonld.gz

function test(filePath, baseIRI = "http://base", fileHash = "hash") {
  const readStream = fs.createReadStream(filePath);
  return readStream.pipe(
    new lib.JsonLdParser({
      baseIRI: baseIRI
    })
  );
}

test(path);
```

```
<--- Last few GCs --->

[21337:0x46742c0]    17677 ms: Mark-sweep 2049.2 (2059.5) -> 2047.6 (2053.3) MB, 634.3 / 0.0 ms  (+ 1022.0 ms in 108 steps since start of marking, biggest step 499.2 ms, walltime since start of marking 1681 ms) (average mu = 0.067, current mu = 0.020) all

<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x13cb519]
    1: StubFrame [pc: 0x1376efc]
Security context: 0x27481be808d1 <JSObject>
    2: slice [0x27481be8e641](this=0x27eb62091cb1 <JSArray[3]>,0,-1)
    3: /* anonymous */ [0x3084b0543681] [/home/melvin/triply/core/jsonld-streaming-parser.js/lib/JsonLdParser.js:~342] [pc=0x56e9b6d4305](this=0x3c246488f741 <Parser map = 0x115fa6e71a39>,0x27eb62090ea9 <Object map = 0x115fa6e73f69>)
    4: /* anonymous */ [0x1e3...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xa07f90 node::Abort() [node]
 2: 0xa0839c node::OnFatalError(char const*, char const*) [node]
 3: 0xb80d9e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xb81119 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xd2d875  [node]
 6: 0xd2df06 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [node]
 7: 0xd3a785 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
 8: 0xd3b635 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0xd3e0ec v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
10: 0xd04cbb v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [node]
11: 0x10464be v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [node]
12: 0x13cb519  [node]
Aborted (core dumped)
```

Note: it also happens when an on data event handler is attached, like so:

```

**Comment by rubensworks** [comment]
Thanks for reporting!

This may be caused by the underlying `jsonparse` dependency.
I think this parser may be keeping to much in memory during parsing.

I'll look into whether or not this can be avoided, otherwise I may have to switch to a different streaming JSON parser (or write my own...).

**Comment by wouterbeek** [comment]
@rubensworks Is it theoretically possible to parse JSON-LD with a streaming JSON parser?  IIUC the format does not guarantee that all transformations are defined up-front.  IOW, at the end of the stream a transformation may be defined that has to be applied to an element in the beginning of the stream.

I'm not sure whether this is a good example, but I'm thinking along the following lines:

```json
{
  "abc": "def",
  "key": "123",
  "@context": {
    "@vocab": "https://example.com/",
    "key": "@id"
  }
}
```

**Comment by rubensworks** [comment]
@wouterbeek Great question, the full answer is available in [this spec](https://www.w3.org/TR/json-ld11-streaming/#streaming-document-form).

TL;DR, under certain assumptions, JSON-LD can be parsed in a streaming manner.
When these assumptions are not met, then the parser can either fail, or fall back to a less efficient mode.
For this, the `streamingProfile` config option of this parser is important, as enabling it will allow the parser to make use of the streaming mode, but it will fail when non-streamable JSON-LD documents are detected.

**Comment by wouterbeek** [comment]
Too bad the order is not enforced in the JSON-LD specification :-(  An alternative would be to parse JSON-LD files twice: once to extract the schema and once to apply it.

**Comment by rubensworks** [comment]
> Too bad the order is not enforced in the JSON-LD specification

True, the only way to enforce this is via the [streaming profile](https://www.w3.org/TR/json-ld11-streaming/#streaming-profile).

> An alternative would be to parse JSON-LD files twice: once to extract the schema and once to apply it.

This is essentially what this parser will do internally when streaming mode is disabled (with some optimizations where possible).
